pkg/engine/forceMutate.go

```diff
@@ -3,112 +3,68 @@ package engine
 		"namespace", resource.GetNamespace(), "name", resource.GetName())
 
+	patchedResource := resource
 	for _, rule := range policy.Spec.Rules {
 		if !rule.HasMutate() {
 			continue
 		}
 
-		rule, err = variables.SubstituteAllForceMutate(log.Log, ctx, rule)
+		r, err := variables.SubstituteAllForceMutate(logger, ctx, removeConditions(rule))
 		if err != nil {
-			return unstructured.Unstructured{}, err
+			return resource, err
 		}
 
-		mutation := rule.Mutation.DeepCopy()
-
-		if mutation.Overlay != nil {
-			overlay := mutation.Overlay
+		if r.Mutation.ForEachMutation != nil {
+			for i, foreach := range r.Mutation.ForEachMutation {
+				patcher := mutate.NewPatcher(r.Name, foreach.PatchStrategicMerge, foreach.PatchesJSON6902, patchedResource, ctx, logger)
+				resp, mutatedResource := patcher.Patch()
+				if resp.Status != response.RuleStatusPass {
+					return patchedResource, fmt.Errorf("foreach mutate result %q at index %d: %s", resp.Status.String(), i, resp.Message)
+				}
 
-			resource, err = mutateResourceWithOverlay(resource, overlay)
-			if err != nil {
-				detailedErr := fmt.Errorf("failed to mutate resource %s with overlay rule %v:%v", resource.GetKind(), rule.Name, err)
-				return unstructured.Unstructured{}, detailedErr
+				patchedResource = mutatedResource
 			}
-		}
-
-		if rule.Mutation.Patches != nil {
-			var resp response.RuleResponse
-			resp, resource = mutate.ProcessPatches(logger.WithValues("rule", rule.Name), rule.Name, rule.Mutation, resource)
+		} else {
+			m := r.Mutation
+			patcher := mutate.NewPatcher(r.Name, m.PatchStrategicMerge, m.PatchesJSON6902, patchedResource, ctx, logger)
+			resp, mutatedResource := patcher.Patch()
 			if resp.Status != response.RuleStatusPass {
-				return unstructured.Unstructured{}, fmt.Errorf(resp.Message)
+				return patchedResource, fmt.Errorf("mutate result %q: %s", resp.Status.String(), resp.Message)
 			}
-		}
 
-		if rule.Mutation.PatchStrategicMerge != nil {
-			var resp response.RuleResponse
-			resp, resource = mutate.ProcessStrategicMergePatch(rule.Name, rule.Mutation.PatchStrategicMerge, resource, logger.WithValues("rule", rule.Name))
-			if resp.Status != response.RuleStatusPass {
-				return unstructured.Unstructured{}, fmt.Errorf(resp.Message)
-			}
+			patchedResource = mutatedResource
 		}
+	}
 
-		if rule.Mutation.PatchesJSON6902 != "" {
-			var resp response.RuleResponse
-			jsonPatches, err := yaml.YAMLToJSON([]byte(rule.Mutation.PatchesJSON6902))
-			if err != nil {
-				return unstructured.Unstructured{}, err
-			}
+	return patchedResource, nil
+}
 
-			resp, resource = mutate.ProcessPatchJSON6902(rule.Name, jsonPatches, resource, logger.WithValues("rule", rule.Name))
-			if resp.Status != response.RuleStatusPass {
-				return unstructured.Unstructured{}, fmt.Errorf(resp.Message)
-			}
+func removeConditions(rule kyverno.Rule) kyverno.Rule {
+	ruleCopy := rule.DeepCopy()
```

Comment:
IMO it's easier to follow if this function works on the rule directly instead of on a deepcopy. That way the deepcopy can just be passed in and handled transparently.

Feels like we do an implicit copy here and then continue working with the copy implicitly which is hard to follow.